docs/audit_report/src/side_channels/01_XX_cmce.rst

```diff
@@ -0,0 +1,182 @@
+       }
+       decode_success &= CT::Mask<uint8_t>(CT::Mask<size_t>::is_equal(hamming_weight_e, sk.params().t()));
+
+       // Check the error vector
+       auto syndrome_from_e = compute_goppa_syndrome(sk.params(), sk.g(), sk.field_ordering(), e);
+       auto syndromes_are_eq = GF_Mask::set();
+       for(size_t i = 0; i < syndrome.size(); ++i) {
+          syndromes_are_eq &= GF_Mask::is_equal(syndrome.at(i), syndrome_from_e.at(i));
+       }
+
+       decode_success &= syndromes_are_eq.elem_mask();
+
+       return {decode_success, std::move(e)};
+    }
+
+Within the `decode()` function the leakage was observed at code lines 102 to 106.
+This is the part where the error vector `e` is obtained.
+The identified leakage was due to an execution flow difference.
+This execution flow difference stems from the code in the `bitvector.h` file at line 223.
+It contains the following constant expression: A bit is set if the given bool is true.
+The relevant line is listed below:
+
+.. code-block::
+
+         private:
+            constexpr bitref& assign(bool bit) noexcept { return (bit) ? set() : unset(); }
+
+When compiled, this line results in a conditional jump instruction in assembly code.
+Depending on the boolean input value a different code branch is executed.
+This expression is executed when assigning values to the elements of `e` at line 104:
+
+.. code-block::
+
+      e.push_back(is_zero_mask.as_bool());
+
+The `push_back()` routine is contained in the file `bitvector.h` at lines 381 to 385 and is implemented as follows:
+
+.. code-block::
+
+      void push_back(bool bit) {
+         const auto i = size();
+         resize(i + 1);
+         ref(i) = bit;
+      }
+
+The `=` operator at line 384 is implemented for the `bitvector` class at line 208 as:
+
+.. code-block::
+
+            constexpr bitref& operator=(bool bit) noexcept { return assign(bit); }
+
+This results in a call of the `assign()` routine listed above.
+
+The identified leakage would allow an adversary to potentially recover the error vector from the code execution, which is security critical.
+
+**Countermeasure implementation and evaluation**
+
+In the following, we propose a countermeasure, implement it and evaluate its effectiveness.
+The underlying issue of the leakage is due to the branch of the `?` operator used in the `assign()` routine.
+This must be avoided, e.g., by performing the set and unset operations regardless of the input value.
+As only one of the two operations is needed at a time, the other operation has to be ineffective.
+Below is a proposal for such a countermeasure:
+
+.. code-block::
+
+         private:
+            constexpr bitref& assign(bool bit) noexcept {
+                const block_type assign_mask = 0 - static_cast<block_type>(bit);
+                this->m_block \|=  (this->m_mask &  assign_mask);
+                this->m_block &= ~(this->m_mask & ~assign_mask);
+                return \*this;
+            }
```

Comment:
We fixed the side channel using Botan's constant time helper class CT::Mask, which is more readable, in my opinion and does similar things as this code block. I guess your code translates better to assembly, though. So I'm fine with you version. (So nothing to do here, just a comment)

Reply:
It's probably worth mentioning that the later (and final) versions of the code fix this side channel

Reply:
Note added in https://github.com/sehlen-bsi/botan-docs/pull/204/commits/5df0aa0d912208888b475bfd854d7198c74bee98